src/coreclr/tools/aot/ILCompiler.Diagnostics/SymNgenWriterWrapper.cs

```diff
@@ -0,0 +1,145 @@
+
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+#nullable enable
+
+namespace Microsoft.DiaSymReader
+{
+    internal class SymNgenWriterWrapper : ISymNGenWriter2, IDisposable
+    {
+        private bool _isDisposed = false;
+        public IntPtr ISymNGenWriter2Inst { get; }
+
+        private SymNgenWriterWrapper(IntPtr writer2Inst)
+        {
+            ISymNGenWriter2Inst = writer2Inst;
+        }
+
+        public static SymNgenWriterWrapper? CreateIfSupported(IntPtr ptr)
+        {
+            var iid = ISymNGenWriter2.IID;
+            int hr = Marshal.QueryInterface(ptr, ref iid, out IntPtr ngenWriterInst);
+            if (hr != 0)
+            {
+                return null;
+            }
+
+            return new SymNgenWriterWrapper(ngenWriterInst);
+        }
+
+        ~SymNgenWriterWrapper()
+        {
+            DisposeInternal();
+        }
+
+        public void Dispose()
+        {
+            DisposeInternal();
+            GC.SuppressFinalize(this);
+        }
+
+        private void DisposeInternal()
+        {
+            if (_isDisposed)
+            {
+                return;
+            }
+            Marshal.Release(ISymNGenWriter2Inst);
+            _isDisposed = true;
+        }
+
+        public unsafe void AddSymbol(string pSymbol, ushort iSection, ulong rva)
+        {
+            IntPtr strLocal = Marshal.StringToBSTR(pSymbol);
+            var inst = ISymNGenWriter2Inst;
+            var func = (delegate* unmanaged<IntPtr, IntPtr, ushort, ulong, int>)(*(*(void***)inst + 3 /* ISymNGenWriter2.AddSymbol slot */));
+            int hr = func(inst, strLocal, iSection, rva);
+            if (hr != 0)
+            {
+                Marshal.FreeBSTR(strLocal);
+                Marshal.ThrowExceptionForHR(hr);
+            }
+        }
+
+        public unsafe void AddSection(ushort iSection, OMF flags, int offset, int cb)
+        {
+            var inst = ISymNGenWriter2Inst;
+            var func = (delegate* unmanaged<IntPtr, ushort, OMF, int, int, int>)(*(*(void***)inst + 4));
+            int hr = func(inst, iSection, flags, offset, cb);
+            if (hr != 0)
+            {
+                Marshal.ThrowExceptionForHR(hr);
+            }
+        }
+
+        public unsafe void OpenModW(string wszModule, string wszObjFile, out UIntPtr ppmod)
+        {
+            IntPtr moduleLocal = Marshal.StringToBSTR(wszModule);
+            IntPtr objLocal = Marshal.StringToBSTR(wszObjFile);
+            var inst = ISymNGenWriter2Inst;
+            var func = (delegate* unmanaged<IntPtr, IntPtr, IntPtr, out UIntPtr, int>)(*(*(void***)inst + 5));
```

Comment:
It's not required, but I'd highly recommend changing this signature to use `UIntPtr*` instead of `out UIntPtr`. Otherwise, you're still producing an IL stub.